graphql/signup.ts

```diff
@@ -0,0 +1,271 @@
+import { syncActorFromAccount } from "@hackerspub/models/actor";
+import { follow } from "@hackerspub/models/following";
+import { createSession } from "@hackerspub/models/session";
+import { USERNAME_REGEXP } from "@hackerspub/models/signin";
+import {
+  createAccount,
+  deleteSignupToken,
+  getSignupToken,
+} from "@hackerspub/models/signup";
+import {
+  generateUuidV7,
+  type Uuid,
+  validateUuid,
+} from "@hackerspub/models/uuid";
+import { getLogger } from "@logtape/logtape";
+import { createGraphQLError } from "graphql-yoga";
+import { Account } from "./account.ts";
+import { builder } from "./builder.ts";
+import { SessionRef } from "./session.ts";
+
+const logger = getLogger(["hackerspub", "graphql", "signup"]);
+
+interface SignupInfo {
+  email: string;
+  inviterId?: Uuid;
+}
+
+interface SignupInput {
+  username: string;
+  name: string;
+  bio: string;
+}
+
+interface SignupValidationErrors {
+  username?: string;
+  name?: string;
+  bio?: string;
+}
+
+const SignupInfoRef = builder.objectRef<SignupInfo>(
+  "SignupInfo",
+);
+
+SignupInfoRef.implement({
+  description:
+    "A signup info containing email and inviter information for account creation.",
+  fields: (t) => ({
+    email: t.exposeString("email"),
+    inviter: t.field({
+      type: Account,
+      nullable: true,
+      async resolve(info, _, ctx) {
+        if (!info.inviterId) return null;
+        const account = await ctx.db.query.accountTable.findFirst({
+          where: { id: info.inviterId },
+          with: { actor: true },
+        });
+        return account || null;
+      },
+    }),
+  }),
+});
+
+const SignupInputRef = builder.inputRef<SignupInput>("SignupInput");
+
+SignupInputRef.implement({
+  description: "Input data for completing account signup.",
+  fields: (t) => ({
+    username: t.string({ required: true }),
+    name: t.string({ required: true }),
+    bio: t.string({ required: true }),
+  }),
+});
+
+const SignupValidationErrorsRef = builder.objectRef<SignupValidationErrors>(
+  "SignupValidationErrors",
+);
+
+SignupValidationErrorsRef.implement({
+  description: "Validation errors for signup fields.",
+  fields: (t) => ({
+    username: t.exposeString("username", { nullable: true }),
+    name: t.exposeString("name", { nullable: true }),
+    bio: t.exposeString("bio", { nullable: true }),
```

Comment:
Instead of representing errors as message strings, how about using [enum types](https://pothos-graphql.dev/docs/guide/enums)? E.g.:

```ts
export const SignupUsernameError = builder.enumType("SignupUsernameError", {
  values: ["MISSING", "TOO_LONG", "INVALID_CHARS", "ALREADY_TAKEN"] as const,
});

export const SignupNameError = builder.enumType("SignupNameError", {
  values: ["MISSING", "TOO_LONG"] as const,
});

export const SignupBioError = builder.enumType("SignupBioError", {
  values: ["TOO_LONG"] as const,
});
```

Since the frontend will need to handle internationalization anyway, I think that would be better.